sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/sources/RatePerMicroBatchStream.scala

```diff
@@ -0,0 +1,155 @@
+    val numRows = limit.asInstanceOf[ReadMaxRows].maxRows()
+
+    val endOffsetLong = Math.min(startOffsetLong + numRows, Long.MaxValue)
+    val endOffset = RatePerMicroBatchStreamOffset(endOffsetLong,
+      timestampAtStartOffset + advanceMsPerBatch)
+
+    endOffset
+  }
+
+  override def deserializeOffset(json: String): Offset = {
+    RatePerMicroBatchStreamOffset.apply(json)
+  }
+
+  override def planInputPartitions(start: Offset, end: Offset): Array[InputPartition] = {
+    val (startOffset, startTimestamp) = extractOffsetAndTimestamp(start)
+    val (endOffset, endTimestamp) = extractOffsetAndTimestamp(end)
+
+    assert(startOffset <= endOffset, s"startOffset($startOffset) > endOffset($endOffset)")
+    assert(startTimestamp <= endTimestamp,
+      s"startTimestamp($startTimestamp) > endTimestamp($endTimestamp)")
+    logDebug(s"startOffset: $startOffset, startTimestamp: $startTimestamp, " +
+      s"endOffset: $endOffset, endTimestamp: $endTimestamp")
+
+    if (startOffset == endOffset) {
+      Array.empty
+    } else {
+      (0 until numPartitions).map { p =>
+        RatePerMicroBatchStreamInputPartition(p, numPartitions, startOffset,
+          startTimestamp, endOffset, endTimestamp)
+      }.toArray
+    }
+  }
+
+  override def createReaderFactory(): PartitionReaderFactory = {
+    RatePerMicroBatchStreamReaderFactory
+  }
+
+  override def commit(end: Offset): Unit = {}
+
+  override def stop(): Unit = {}
+
+  override def toString: String = s"RatePerMicroBatchStream[rowsPerBatch=$rowsPerBatch, " +
+    s"numPartitions=$numPartitions, startTimestamp=$startTimestamp, " +
+    s"advanceMsPerBatch=$advanceMsPerBatch]"
+}
+
+case class RatePerMicroBatchStreamOffset(offset: Long, timestamp: Long) extends Offset {
+  override def json(): String = {
+    Serialization.write(this)(RatePerMicroBatchStreamOffset.formats)
+  }
+}
+
+object RatePerMicroBatchStreamOffset {
+  implicit val formats = Serialization.formats(NoTypeHints)
+
+  def apply(json: String): RatePerMicroBatchStreamOffset =
+    Serialization.read[RatePerMicroBatchStreamOffset](json)
+}
+
+case class RatePerMicroBatchStreamInputPartition(
+    partitionId: Int,
+    numPartitions: Int,
+    startOffset: Long,
+    startTimestamp: Long,
+    endOffset: Long,
+    endTimestamp: Long) extends InputPartition
+
+object RatePerMicroBatchStreamReaderFactory extends PartitionReaderFactory {
+  override def createReader(partition: InputPartition): PartitionReader[InternalRow] = {
+    val p = partition.asInstanceOf[RatePerMicroBatchStreamInputPartition]
+    new RatePerMicroBatchStreamPartitionReader(p.partitionId, p.numPartitions,
+      p.startOffset, p.startTimestamp, p.endOffset, p.endTimestamp)
+  }
+}
+
+class RatePerMicroBatchStreamPartitionReader(
+    partitionId: Int,
+    numPartitions: Int,
+    startOffset: Long,
+    startTimestamp: Long,
+    endOffset: Long,
+    endTimestamp: Long) extends PartitionReader[InternalRow] {
```

Comment:
`endTimestamp` is useless in `RatePerMicroBatchStreamPartitionReader`. All output rows have the same timestamp value (`startTimestamp`)?

Reply:
Yes, actually I thought about gradual increasing of timestamp even in same micro-batch, but decided to make it simple and add if someone claims with good use cases. I'll remove the unused parameter for now.

Reply:
Ok.